### PR TITLE
Add audit model

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description 
This adds the first version of an audit model.

Admin
image

Database

image

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -6,10 +6,63 @@ Django eduNEXT Audit Model
...
+
+1. Status. If the process was successful or not.
+2. Action. The string given to identify the process.
+3. Time stamp. The execute date.
+4. Method name. Method or function name.
+5. Traceback log. If there was an exception, this will contain the traceback.
+6. Site domain. The domain of the current site.
+7. Performer. The user who started the method, this depend on the request.user
+8. Input. The values used to execute the method.
+9. Output. The value returned by the method.
```
Comment (felipemontoya) [inline README.rst:43]: Should we store the IP of the incoming request?

File: README.rst [hunk 2]
```diff
@@ -6,10 +6,63 @@ Django eduNEXT Audit Model
...
+
 Usage
 #####
+Audit any execution of a method or function. This will create a database register with the following information:
+
+1. Status. If the process was successful or not.
+2. Action. The string given to identify the process.
+3. Time stamp. The execute date.
+4. Method name. Method or function name.
+5. Traceback log. If there was an exception, this will contain the traceback.
```
Comment (felipemontoya) [inline README.rst:39]: Does this have to be a traceback log? I mean could we log the stdout when the status is success?

File: eox_audit_model/models.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        method_name: Audited method.
+        output_parameters: Values returned by the audited method.
+        performer: Username of the user who is executing the current request.
+        site_domain: Domain get from the current request.
+        status: Success or fail. that depends on audited method.
+
+    Methods:
+        execute_action: Allow to audit the execution for the given method.
+    """
+
```
Comment (felipemontoya) [inline eox_audit_model/models.py:65]: should we worry about leaking the consecutive ID?
We could also also add a UUID that could act as an external key. Thoughts on this?

File: eox_audit_model/models.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    action = models.CharField(max_length=150)
+    audit_date_stamp = models.DateTimeField(auto_now_add=True, editable=False)
+    input_parameters = JSONField(blank=True)
+    traceback_log = models.TextField(null=True, blank=True)
+    method_name = models.CharField(max_length=150)
+    output_parameters = JSONField(blank=True)
+    performer = models.CharField(max_length=150, default=get_current_performer)
+    site_domain = models.CharField(max_length=150, default=get_current_domain)
```
Comment (felipemontoya) [inline eox_audit_model/models.py:73]: Why 150? a full domain can contain up to 253 chars.

File: eox_audit_model/models.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        execute_action: Allow to audit the execution for the given method.
+    """
+
+    action = models.CharField(max_length=150)
+    audit_date_stamp = models.DateTimeField(auto_now_add=True, editable=False)
+    input_parameters = JSONField(blank=True)
+    traceback_log = models.TextField(null=True, blank=True)
+    method_name = models.CharField(max_length=150)
+    output_parameters = JSONField(blank=True)
+    performer = models.CharField(max_length=150, default=get_current_performer)
```
Comment (felipemontoya) [inline eox_audit_model/models.py:72]: are we  going to store the "username" or the user_id? is this a foreing key or just a text reference?

File: eox_audit_model/models.py (hunk 4 above)
Comment (felipemontoya) [reply eox_audit_model/models.py:73]: Another question would be, should we log the domain, or use the django_sites model instead?

Comment (andrey-canon) [reply eox_audit_model/models.py:73]: edunext doesn't use sites for all the tenants

File: eox_audit_model/models.py (hunk 5 above)
Comment (andrey-canon) [reply eox_audit_model/models.py:72]: `username` and this is a text reference

**Comment by andrey-canon** [comment]
> I like many of the ideas here. This proposes that we use only one table (model) to cover all and any audit trails, correct?
> 
> Now, there are two use cases we are not covering and I would like to at least discuss.
> 
> * should we create a decorator that we can put on the view and that it takes care of everything?
> * what if I want to audit an action that is not a method of a class. Should this model expose some functions to do just that?

This uses two models audit for the general audit process and auidnotes  if required information has to be stored.
I like the decorator idea.
This supports functions and class methods, what is different action in this case ????

File: README.rst (hunk 2 above)
Comment (andrey-canon) [reply README.rst:39]: do you mean the normal django logs ?

File: eox_audit_model/models.py (hunk 3 above)
Comment (andrey-canon) [reply eox_audit_model/models.py:65]: I don't see any advantages ...

File: README.rst (hunk 2 above)
Comment (felipemontoya) [reply README.rst:39]: yes, exactly. That or the output of `print()`

File: eox_audit_model/models.py (hunk 3 above)
Comment (felipemontoya) [reply eox_audit_model/models.py:65]: https://stackoverflow.com/a/396581 point 3

**Pull request merged**